core/src/main/java/com/epam/esm/dao/impl/GiftDaoImpl.java

```diff
@@ -0,0 +1,282 @@
+
+
+    private static final String SORT_BY_CREATE_DATE_ASC = " order by create_date asc";
+    private static final String SORT_BY_CREATE_DATE_DESC = " order by create_date desc";
+    private static final String SORT_BY_NAME_ASC = " order by name asc";
+    private static final String SORT_BY_NAME_DESC = " order by name desc";
+    private static final String DELETE_GIFT_BY_ID = "delete from gift_certificate where id = ?";
+    private static final String UPDATE_BY_ID_QUERY = "UPDATE gift_certificate set name=?, description=?, price=?, duration=? WHERE id=?;";
+    private static final String SELECT_TAGS_BY_GIFT = "SELECT t.id, t.name "
+            + "FROM gift_tags gt, tag t "
+            + "WHERE gt.tag_id = t.id and gt.gift_id=?;";
+    private static final String INSERT_GIFT_QUERY = "insert into gift_certificate (name,description,price,duration,create_date,last_update_date) values (?, ?, ?, ?, ?, ?)";
+
+    private static final String ZERO_OR_MORE_ELEMENTS_WILDCARD = "%";
+
+    private static final String COLUMN_ID = "id";
+    private static final String COLUMN_NAME = "name";
+    private static final String COLUMN_DESCRIPTION = "description";
+    private static final String COLUMN_PRICE = "price";
+    private static final String COLUMN_DURATION = "duration";
+    private static final String COLUMN_CREATE_DATE = "create_date";
+    private static final String COLUMN_LAST_UPDATE_DATE = "last_update_date";
+
+
+
+
+    RowMapper<GiftCertificateEntity> GIFT_ROW_MAPPER = (ResultSet resultSet, int rowNum) -> {
+        return new GiftCertificateEntity(
+                resultSet.getLong(COLUMN_ID),
+                resultSet.getString(COLUMN_NAME),
+                resultSet.getString(COLUMN_DESCRIPTION),
+                resultSet.getInt(COLUMN_PRICE),
+                resultSet.getInt(COLUMN_DURATION),
+                resultSet.getTimestamp(COLUMN_CREATE_DATE),
+                resultSet.getTimestamp(COLUMN_LAST_UPDATE_DATE));
+    };
+
+    RowMapper<TagEntity> TAG_ROW_MAPPER = (ResultSet resultSet, int rowNum) -> {
+        return new TagEntity(
+                resultSet.getLong(COLUMN_ID),
+                resultSet.getString(COLUMN_NAME));
+    };
+
+    @Autowired
+    public GiftDaoImpl(JdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+
+    @Override
+    public List<GiftCertificateEntity> findAllGifts(){
+        List<GiftCertificateEntity> query = jdbcTemplate.query(FIND_ALL_GIFTS_QUERY, GIFT_ROW_MAPPER);
+
+        for (GiftCertificateEntity giftCertificateEntity : query){
+            Set<TagEntity> tags = getTagsByGiftId(giftCertificateEntity.getId());
+            giftCertificateEntity.setTags(tags);
+        }
+
+        return query;
+    }
+
+    @Override
+    public GiftCertificateEntity findGiftById(Long giftId){
+        List<GiftCertificateEntity> query = jdbcTemplate.query(FIND_GIFT_BY_ID_QUERY, GIFT_ROW_MAPPER, giftId);
```

Comment:
you need to get tags when you are getting gift
please use join for this purpose
you must have only one call to db for getting gift and all its tags

Reply:
this is one call to db for get all tags for one gift certificate